File: sdk/python/src/openlit/guard/pii.py
"""
PII & secret detection guard.

Uses ~25 high-confidence regex patterns for API keys, PII, and secrets.
Runs locally in <1 ms.  Supports ``redact``, ``deny``, and ``warn`` actions.
Phases: preflight + postflight.
"""

from __future__ import annotations

import re
from typing import Dict, List, Optional, Tuple

from openlit.guard._base import Guard, GuardAction, GuardPhase, GuardResult

# Each pattern is (label, compiled_regex).  Labels are used in
# ``[REDACTED:<label>]`` placeholders and in ``GuardResult.classification``.
_PATTERNS: List[Tuple[str, re.Pattern]] = []


def _p(label: str, pattern: str, flags: int = 0) -> None:
    _PATTERNS.append((label, re.compile(pattern, flags)))


# ---- API keys / tokens ----
_p("openai-api-key", r"sk-(?:proj-)?[A-Za-z0-9_-]{20,}")
_p("anthropic-api-key", r"sk-ant-[A-Za-z0-9_-]{20,}")
_p("aws-access-key", r"(?<![A-Za-z0-9/+=])AKIA[0-9A-Z]{16}(?![A-Za-z0-9/+=])")
_p("gcp-api-key", r"AIza[0-9A-Za-z_-]{35}")
_p("github-token", r"(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9_]{36,}")
_p("github-fine-grained", r"github_pat_[A-Za-z0-9_]{22,}")
_p("stripe-key", r"(?:sk|pk)_(?:live|test)_[A-Za-z0-9]{24,}")
_p("slack-token", r"xox[bpoas]-[A-Za-z0-9-]{10,}")
_p(
    "slack-webhook",
    r"https://hooks\.slack\.com/services/T[A-Za-z0-9]+/B[A-Za-z0-9]+/[A-Za-z0-9]+",
)
_p("twilio-api-key", r"SK[0-9a-fA-F]{32}")
_p("sendgrid-api-key", r"SG\.[A-Za-z0-9_-]{22}\.[A-Za-z0-9_-]{43}")
_p("mailgun-api-key", r"key-[0-9a-zA-Z]{32}")
_p("azure-key", r"[0-9a-f]{32}")  # very generic; kept short to reduce FPs
_p("heroku-api-key", r"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}")

# ---- PII ----
_p("email", r"[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z]{2,}")
_p("phone-us", r"(?<!\d)(?:\+?1[-.\s]?)?(?:\(?\d{3}\)?[-.\s]?)\d{3}[-.\s]?\d{4}(?!\d)")
_p("ssn", r"(?<!\d)\d{3}-\d{2}-\d{4}(?!\d)")
_p(
    "credit-card",
    r"(?<!\d)(?:4\d{3}|5[1-5]\d{2}|3[47]\d{2}|6(?:011|5\d{2}))[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}(?!\d)",
)
_p(
    "ipv4",
    r"(?<!\d)(?:25[0-5]|2[0-4]\d|[01]?\d\d?)(?:\.(?:25[0-5]|2[0-4]\d|[01]?\d\d?)){3}(?!\d)",
)

# ---- Secrets / credentials ----
_p("bearer-token", r"Bearer\s+[A-Za-z0-9_\-.~+/]+=*", re.IGNORECASE)
_p("basic-auth", r"Basic\s+[A-Za-z0-9+/]+=*", re.IGNORECASE)
_p("private-key", r"-----BEGIN (?:RSA |EC |DSA |OPENSSH )?PRIVATE KEY-----")
_p("connection-string", r"(?:mongodb(?:\+srv)?|postgres(?:ql)?|mysql|redis|amqp)://\S+")
_p("env-secret", r"(?i)(?:password|secret|token|api_key|apikey)\s*[=:]\s*\S+")


class PII(Guard):
    """
    Detects PII and secrets using high-confidence regex patterns.

    Parameters
    ----------
    action : str
        ``"redact"`` (replace matches with ``[REDACTED:label]``),
        ``"deny"`` (block entirely), or ``"warn"`` (emit event only).
    custom_patterns : dict, optional
        Extra ``{label: regex_string}`` patterns to add.
    """

    name = "pii"
    phases = (GuardPhase.PREFLIGHT, GuardPhase.POSTFLIGHT)

    def __init__(
        self,
        action: str = "redact",
        custom_patterns: Optional[Dict[str, str]] = None,
        **kwargs,
    ):
        super().__init__(action=action, **kwargs)
        self._extra: List[Tuple[str, re.Pattern]] = []
        if custom_patterns:
            for label, pat in custom_patterns.items():
                self._extra.append((label, re.compile(pat)))

    def evaluate(self, text: str) -> GuardResult:
        all_patterns = _PATTERNS + self._extra
        matches: List[Tuple[str, re.Match]] = []

        for label, pattern in all_patterns:
            for m in pattern.finditer(text):
                matches.append((label, m))

        if not matches:
            return GuardResult(guard_name=self.name)

        labels = sorted({label for label, _ in matches})
        classification = ", ".join(labels)
        best_score = min(1.0, len(matches) * 0.2 + 0.5)

        transformed: Optional[str] = None
        if self._action == GuardAction.REDACT:
            sorted_matches = sorted(matches, key=lambda x: x[1].start(), reverse=True)
            result_text = text
            for label, m in sorted_matches:
                result_text = (
                    result_text[: m.start()]
                    + f"[REDACTED:{label}]"
                    + result_text[m.end() :]
                )
            transformed = result_text

        return GuardResult(
            action=self._action,
            score=round(best_score, 2),
            guard_name=self.name,
            classification=classification,
            explanation=f"Detected {len(matches)} PII/secret match(es): {classification}",
            transformed_text=transformed,
        )

File: sdk/typescript/src/guard/pii.ts
/**
 * PII guard -- detects and optionally redacts personally identifiable
 * information, API keys, and secrets using regex patterns.
 *
 * Patterns must stay in sync with: sdk/python/src/openlit/guard/pii.py
 */

import { Guard, GuardAction, GuardPhase, GuardResult, makeGuardResult, GuardOptions } from './base';

interface PatternEntry {
  label: string;
  regex: RegExp;
}

// ---- API keys / tokens ----
const PATTERNS: PatternEntry[] = [
  { label: 'openai-api-key', regex: /sk-(?:proj-)?[A-Za-z0-9_-]{20,}/ },
  { label: 'anthropic-api-key', regex: /sk-ant-[A-Za-z0-9_-]{20,}/ },
  { label: 'aws-access-key', regex: /(?<![A-Za-z0-9/+=])AKIA[0-9A-Z]{16}(?![A-Za-z0-9/+=])/ },
  { label: 'gcp-api-key', regex: /AIza[0-9A-Za-z_-]{35}/ },
  { label: 'github-token', regex: /(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9_]{36,}/ },
  { label: 'github-fine-grained', regex: /github_pat_[A-Za-z0-9_]{22,}/ },
  { label: 'stripe-key', regex: /(?:sk|pk)_(?:live|test)_[A-Za-z0-9]{24,}/ },
  { label: 'slack-token', regex: /xox[bpoas]-[A-Za-z0-9-]{10,}/ },
  { label: 'slack-webhook', regex: /https:\/\/hooks\.slack\.com\/services\/T[A-Za-z0-9]+\/B[A-Za-z0-9]+\/[A-Za-z0-9]+/ },
  { label: 'twilio-api-key', regex: /SK[0-9a-fA-F]{32}/ },
  { label: 'sendgrid-api-key', regex: /SG\.[A-Za-z0-9_-]{22}\.[A-Za-z0-9_-]{43}/ },
  { label: 'mailgun-api-key', regex: /key-[0-9a-zA-Z]{32}/ },
  { label: 'azure-key', regex: /[0-9a-f]{32}/ },
  { label: 'heroku-api-key', regex: /[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/ },
  // ---- PII ----
  { label: 'email', regex: /[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z]{2,}/ },
  { label: 'phone-us', regex: /(?<!\d)(?:\+?1[-.\s]?)?(?:\(?\d{3}\)?[-.\s]?)\d{3}[-.\s]?\d{4}(?!\d)/ },
  { label: 'ssn', regex: /(?<!\d)\d{3}-\d{2}-\d{4}(?!\d)/ },
  { label: 'credit-card', regex: /(?<!\d)(?:4\d{3}|5[1-5]\d{2}|3[47]\d{2}|6(?:011|5\d{2}))[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}(?!\d)/ },
  { label: 'ipv4', regex: /(?<!\d)(?:25[0-5]|2[0-4]\d|[01]?\d\d?)(?:\.(?:25[0-5]|2[0-4]\d|[01]?\d\d?)){3}(?!\d)/ },
  // ---- Secrets / credentials ----
  { label: 'bearer-token', regex: /Bearer\s+[A-Za-z0-9_\-.~+/]+=*/i },
  { label: 'basic-auth', regex: /Basic\s+[A-Za-z0-9+/]+=*/i },
  { label: 'private-key', regex: /-----BEGIN (?:RSA |EC |DSA |OPENSSH )?PRIVATE KEY-----/ },
  { label: 'connection-string', regex: /(?:mongodb(?:\+srv)?|postgres(?:ql)?|mysql|redis|amqp):\/\/\S+/ },
  { label: 'env-secret', regex: /(?:password|secret|token|api_key|apikey)\s*[=:]\s*\S+/i },
];

export interface PIIOptions extends GuardOptions {
  customPatterns?: Record<string, string>;
}

export class PII extends Guard {
  readonly name = 'pii';
  readonly phases = [GuardPhase.PREFLIGHT, GuardPhase.POSTFLIGHT];

  private readonly _allPatterns: PatternEntry[];

  constructor(opts: PIIOptions = {}) {
    super({ action: opts.action ?? 'redact', maxScanLength: opts.maxScanLength });
    const extra: PatternEntry[] = [];
    if (opts.customPatterns) {
      for (const [label, pat] of Object.entries(opts.customPatterns)) {
        extra.push({ label, regex: new RegExp(pat, 'gi') });
      }
    }
    this._allPatterns = [
      ...PATTERNS.map(({ label, regex }) => ({
        label,
        regex: new RegExp(regex.source, regex.flags.includes('g') ? regex.flags : regex.flags + 'g'),
      })),
      ...extra,
    ];
  }

  evaluate(text: string): GuardResult {
    const matches: Array<{ label: string; start: number; end: number }> = [];

    for (const { label, regex } of this._allPatterns) {
      regex.lastIndex = 0;
      let m: RegExpExecArray | null;
      while ((m = regex.exec(text)) !== null) {
        matches.push({ label, start: m.index, end: m.index + m[0].length });
      }
    }

    if (matches.length === 0) {
      return makeGuardResult({ guardName: this.name });
    }

    const labels = [...new Set(matches.map((m) => m.label))].sort();
    const classification = labels.join(', ');
    const bestScore = Math.min(1.0, matches.length * 0.2 + 0.5);

    let transformedText: string | null = null;
    if (this._action === GuardAction.REDACT) {
      const sortedMatches = [...matches].sort((a, b) => b.start - a.start);
      let resultText = text;
      for (const { label, start, end } of sortedMatches) {
        resultText = resultText.slice(0, start) + `[REDACTED:${label}]` + resultText.slice(end);
      }
      transformedText = resultText;
    }

    return makeGuardResult({
      action: this._action,
      score: Math.round(bestScore * 100) / 100,
      guardName: this.name,
      classification,
      explanation: `Detected ${matches.length} PII/secret match(es): ${classification}`,
      transformedText,
    });
  }
}
